File: src/test/java/net/loxal/soa/restkit/endpoint/ballot/PollResourceIT.java
/*
 * Copyright 2014 Alexander Orlov <alexander.orlov@loxal.net>. All rights reserved.
 */

package net.loxal.soa.restkit.endpoint.ballot;

import net.loxal.soa.restkit.endpoint.AbstractEndpointTest;
import net.loxal.soa.restkit.endpoint.Endpoint;
import net.loxal.soa.restkit.model.ballot.Poll;
import net.loxal.soa.restkit.model.common.ErrorMessage;
import org.junit.Test;

import javax.ws.rs.client.Entity;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import static org.junit.Assert.assertEquals;

public class PollResourceIT extends AbstractEndpointTest {
    private static final String POLL_QUESTION = "What is the meaning of life?";
    private static final List<String> POLL_ANSWERS = Arrays.asList("Yes", "No");

    static Response createPoll() {
        Poll poll = new Poll(POLL_QUESTION, POLL_ANSWERS);

        Response createdPoll = prepareGenericRequest(PollResource.RESOURCE_PATH).request().post(Entity.json(poll));
        assertEquals(Response.Status.CREATED.getStatusCode(), createdPoll.getStatus());
        assertEquals(MediaType.APPLICATION_JSON_TYPE, createdPoll.getMediaType());

        return createdPoll;
    }

    @Test
    public void createNewPoll() throws Exception {
        Response response = createPoll();

        assertEquals(Response.Status.CREATED.getStatusCode(), response.getStatus());
        assertEquals(MediaType.APPLICATION_JSON_TYPE, response.getMediaType());
        assertEquals(false, response.getLocation().getPath().endsWith("null"));
        assertEquals(true, response.getLocation().getPath().startsWith(Endpoint.URI_PATH_SEPARATOR + PollResource.RESOURCE_PATH));
        assertEquals(true, response.getLocation().getSchemeSpecificPart().contains(PollResource.RESOURCE_PATH + Endpoint.URI_PATH_SEPARATOR));
        assertEquals(false, response.getLocation().getSchemeSpecificPart().endsWith(PollResource.RESOURCE_PATH));
    }

    @Test
    public void deleteNonExistentPoll() throws Exception {
        Response existingPoll = createPoll();

        Response response = prepareTarget(existingPoll.getLocation() + NON_EXISTENT).request().delete();

        assertEquals(MediaType.APPLICATION_JSON_TYPE, response.getMediaType());
        Response.Status notFoundStatus = Response.Status.NOT_FOUND;
        assertEquals(notFoundStatus.getStatusCode(), response.getStatus());
        ErrorMessage errorMessage = response.readEntity(ErrorMessage.class);
        assertEquals(Response.Status.BAD_REQUEST.getReasonPhrase(), errorMessage.getType());
    }

    @Test
    public void deleteExistingPoll() throws Exception {
        Response existingPoll = createPoll();

        Response deletion = prepareTarget(existingPoll.getLocation()).request().delete();

        assertEquals(Response.Status.NO_CONTENT.getStatusCode(), deletion.getStatus());
        assertEquals(MediaType.APPLICATION_JSON_TYPE, deletion.getMediaType());
    }

    @Test
    public void retrieveExistingPoll() throws Exception {
        Response existingPoll = createPoll();

        Response retrieval = prepareTarget(existingPoll.getLocation()).request().get();

        assertEquals(Response.Status.OK.getStatusCode(), retrieval.getStatus());
        assertEquals(MediaType.APPLICATION_JSON_TYPE, retrieval.getMediaType());
        Poll retrievedPoll = retrieval.readEntity(Poll.class);
        assertEquals(POLL_QUESTION, retrievedPoll.getQuestion());
        assertEquals(POLL_ANSWERS, retrievedPoll.getAnswers());
        assertEquals(2, retrievedPoll.getAnswers().size());
        assertEquals("Yes", retrievedPoll.getAnswers().get(0));
        assertEquals("No", retrievedPoll.getAnswers().get(1));
    }

    @Test
    public void retrieveNonExistentPoll() throws Exception {
        Response existingPoll = createPoll();

        Response retrieval = prepareTarget(existingPoll.getLocation() + NON_EXISTENT).request().get();


        assertEquals(Response.Status.NOT_FOUND.getStatusCode(), retrieval.getStatus());
        assertEquals(MediaType.APPLICATION_JSON_TYPE, retrieval.getMediaType());
        ErrorMessage notFoundError = retrieval.readEntity(ErrorMessage.class);
        assertEquals(Response.Status.BAD_REQUEST.getReasonPhrase(), notFoundError.getType());
    }

    @Test
    public void updateExistingPoll() throws Exception {
        Response existingPoll = createPoll();

        String newQuestion = "Does it work?";
        String firstAnswerOption = "Maybe";
        String secondAnswerOption = "Somewhat";
        String thirdAnswerOption = "Absolutely Not";
        List<String> newAnswerOptions = Arrays.asList(firstAnswerOption, secondAnswerOption, thirdAnswerOption);

        Poll modifiedPoll = new Poll(newQuestion, newAnswerOptions);

        Response update = prepareTarget(existingPoll.getLocation()).request().put(Entity.json(modifiedPoll));

        assertEquals(Response.Status.OK.getStatusCode(), update.getStatus());
        assertEquals(MediaType.APPLICATION_JSON_TYPE, update.getMediaType());

        Response retrievedUpdatedPoll = prepareTarget(existingPoll.getLocation()).request().get();
        Poll updatedPoll = retrievedUpdatedPoll.readEntity(Poll.class);
        assertEquals(newQuestion, updatedPoll.getQuestion());
        assertEquals(newAnswerOptions.size(), updatedPoll.getAnswers().size());
        assertEquals(firstAnswerOption, updatedPoll.getAnswers().get(0));
        assertEquals(secondAnswerOption, updatedPoll.getAnswers().get(1));
        assertEquals(thirdAnswerOption, updatedPoll.getAnswers().get(2));
    }

    @Test
    public void updateNonExistentPoll() throws Exception {
        Response existingPoll = createPoll();

        Poll somePoll = new Poll("Irrelevant", Collections.emptyList());
        Response update = prepareTarget(existingPoll.getLocation() + NON_EXISTENT).request().put(Entity.json(somePoll));

        assertEquals(Response.Status.NOT_FOUND.getStatusCode(), update.getStatus());
        assertEquals(MediaType.APPLICATION_JSON_TYPE, update.getMediaType());
    }
}

File: src/test/kotlin/net/loxal/soa/restkit/endpoint/ballot/PollResourceIT.kt
/*
 * Copyright 2015 Alexander Orlov <alexander.orlov@loxal.net>. All rights reserved.
 */

package net.loxal.soa.restkit.endpoint.ballot

import net.loxal.soa.restkit.endpoint.AbstractEndpointTest
import net.loxal.soa.restkit.endpoint.Endpoint
import net.loxal.soa.restkit.model.ballot.Poll
import net.loxal.soa.restkit.model.common.ErrorMessage
import org.junit.Test
import java.util.*
import javax.ws.rs.client.Entity
import javax.ws.rs.core.MediaType
import javax.ws.rs.core.Response
import kotlin.test.assertEquals

class PollResourceIT : AbstractEndpointTest() {
    @Test
    fun createNewPoll() {
        val response = createEntity()

        assertEquals(Response.Status.CREATED.statusCode, response.status)
        assertEquals(MediaType.APPLICATION_JSON_TYPE, response.mediaType)
        assertEquals(false, response.location.path.endsWith("null"))
        assertEquals(true, response.location.path.startsWith(Endpoint.URI_PATH_SEPARATOR + PollResource.RESOURCE_PATH))
        assertEquals(true, response.location.schemeSpecificPart.contains(PollResource.RESOURCE_PATH + Endpoint.URI_PATH_SEPARATOR))
        assertEquals(false, response.location.schemeSpecificPart.endsWith(PollResource.RESOURCE_PATH))
    }

    @Test
    public fun deleteNonExistentPoll() {
        val existingEntity = createEntity()

        val response = AbstractEndpointTest.prepareTarget("${existingEntity.location} ${AbstractEndpointTest.NON_EXISTENT}").request().delete()

        assertEquals(MediaType.APPLICATION_JSON_TYPE, response.mediaType)
        val notFoundStatus = Response.Status.NOT_FOUND
        assertEquals(notFoundStatus.statusCode, response.status)
        val errorMessage = response.readEntity(ErrorMessage::class.java)
        assertEquals(Response.Status.BAD_REQUEST.reasonPhrase, errorMessage.reasonPhrase)
    }

    @Test
    public fun deleteExistingPoll() {
        val existingPoll = createEntity()

        val deletion = AbstractEndpointTest.prepareTarget(existingPoll.location).request().delete()

        assertEquals(Response.Status.NO_CONTENT.statusCode, deletion.status)
        assertEquals(null, deletion.mediaType)
    }

    @Test
    public fun retrieveExistingPoll() {
        val existingPoll = createEntity()

        val retrieval = AbstractEndpointTest.prepareTarget(existingPoll.location).request().get()

        assertEquals(Response.Status.OK.statusCode, retrieval.status)
        assertEquals(MediaType.APPLICATION_JSON_TYPE, retrieval.mediaType)
        val retrievedPoll = retrieval.readEntity<net.loxal.soa.restkit.model.ballot.Poll>(Poll::class.java)
        assertEquals(POLL_QUESTION, retrievedPoll.question)
        assertEquals(POLL_OPTIONS, retrievedPoll.options)
        assertEquals(true, retrievedPoll.multipleAnswers)
        assertEquals(CORRECT_ANSWERS, retrievedPoll.correctAnswers)
        assertEquals(2, retrievedPoll.options.size)
        assertEquals("Yes", retrievedPoll.options[0])
        assertEquals("No", retrievedPoll.options[1])
    }

    @Test
    public fun retrieveNonExistentPoll() {
        val existingEntity = createEntity()

        val retrieval = AbstractEndpointTest.prepareTarget("${existingEntity.location} ${AbstractEndpointTest.NON_EXISTENT}").request().get()


        assertEquals(Response.Status.NOT_FOUND.statusCode, retrieval.status)
        assertEquals(MediaType.APPLICATION_JSON_TYPE, retrieval.mediaType)
        val notFoundError = retrieval.readEntity(ErrorMessage::class.java)
        assertEquals(Response.Status.BAD_REQUEST.reasonPhrase, notFoundError.reasonPhrase)
    }

    @Test
    public fun updateExistingPoll() {
        val existingPoll = createEntity()

        val newQuestion = "Does it work?"
        val firstAnswerOption = "Maybe"
        val secondAnswerOption = "Somewhat"
        val thirdAnswerOption = "Absolutely Not"
        val newAnswerOptions = Arrays.asList<String>(firstAnswerOption, secondAnswerOption, thirdAnswerOption)

        val modifiedPoll = Poll(newQuestion, newAnswerOptions)

        val update = AbstractEndpointTest.prepareTarget(existingPoll.location).request().put(Entity.json<Poll>(modifiedPoll))

        assertEquals(Response.Status.OK.statusCode, update.status)
        assertEquals(MediaType.APPLICATION_JSON_TYPE, update.mediaType)

        val retrievedUpdatedPoll = AbstractEndpointTest.prepareTarget(existingPoll.location).request().get()
        val updatedPoll = retrievedUpdatedPoll.readEntity<net.loxal.soa.restkit.model.ballot.Poll>(Poll::class.java)
        assertEquals(newQuestion, updatedPoll.question)
        assertEquals(newAnswerOptions.size.toInt(), updatedPoll.options.size.toInt())
        assertEquals(firstAnswerOption, updatedPoll.options[0])
        assertEquals(secondAnswerOption, updatedPoll.options[1])
        assertEquals(thirdAnswerOption, updatedPoll.options[2])
    }

    @Test
    public fun updateNonExistentPoll() {
        val existingEntity = createEntity()

        val somePoll = Poll("Irrelevant", listOf<String>())
        val update = AbstractEndpointTest.prepareTarget("${existingEntity.location} ${AbstractEndpointTest.NON_EXISTENT}").request().put(Entity.json<Poll>(somePoll))

        assertEquals(Response.Status.NOT_FOUND.statusCode, update.status)
        assertEquals(MediaType.APPLICATION_JSON_TYPE, update.mediaType)
    }

    companion object {
        private val POLL_QUESTION = "What is the meaning of life?"
        private val POLL_OPTIONS = listOf("Yes", "No")
        private val CORRECT_ANSWERS = listOf(1)

        fun createEntity(
                id: String = UUID.randomUUID().toString(),
                poll: Poll = Poll(POLL_QUESTION, options = POLL_OPTIONS, correctAnswers = CORRECT_ANSWERS, multipleAnswers = true)
        ): Response {
            val createdPoll = AbstractEndpointTest.prepareGenericRequest(PollResource.RESOURCE_PATH)
                    .path(id).request()
                    .post(Entity.json<Poll>(poll))
            assertEquals(Response.Status.CREATED.statusCode, createdPoll.status)
            assertEquals(MediaType.APPLICATION_JSON_TYPE, createdPoll.mediaType)

            return createdPoll
        }
    }
}